File: app-serv/internal/schema/usage_live.go
// Package schema holds request/response DTOs and their validation rules.
//
// @file      internal/schema/usage_live.go
// @for       The live Usage frame: the three facts a period window cannot carry,
//
//	in the shape the panel parses.
//
// @uses      internal/domain, time.
// @reason    SPEC-UI-001 §6.5 fixes the frame's contents (the requests in flight
//
//	now, the requests that just finished, and the provider the gateway
//	last reported an error for) and app-ui already parses them with a Zod
//	schema, so the member names are a contract between two applications
//	rather than a local choice. Keeping the mapping here means the
//	handler writes bytes and nothing else, and the panel's own rule (a
//	key the frame omits is not a statement that nothing is happening)
//	is satisfied by construction: every member is always present.
//
//	Both lists are always rendered as arrays, never as `null`. Go marshals
//	a nil slice as `null`, and the panel normalizes that, but a frame
//	that says "nothing is running" and a frame that says nothing are
//	different facts to a reader diffing the wire, so the empty case is
//	spelled one way.
//
// @author    Dodi Rusmana <rusmanadodi@kentangtech.com>
// @layer     schema
// @stability experimental
// @since     2026-09-22
package schema

import (
	"github.com/rusmanadodi2598/pannelAI/app-serv/internal/domain"
)

// UsageLiveActive is one request the gateway is routing right now.
//
// `started_at` is present because the panel's staleness guard reads it: without
// it the panel could not tell a request still running from one whose marker the
// gateway never cleared, and it would light a node forever.
type UsageLiveActive struct {
	ProviderID string `json:"provider_id"`
	EndpointID string `json:"endpoint_id"`
	Model      string `json:"model"`
	StartedAt  string `json:"started_at"`
}

// UsageLiveRecent is one request that finished recently. The list is bounded by
// the read, so the panel renders it rather than paging it.
type UsageLiveRecent struct {
	RequestID  string `json:"request_id"`
	ProviderID string `json:"provider_id"`
	Model      string `json:"model"`
	TS         string `json:"ts"`
	Status     string `json:"status"`
	TokensIn   int64  `json:"tokens_in"`
	TokensOut  int64  `json:"tokens_out"`
	ErrorCode  string `json:"error_code"`
}

// UsageLiveFrame is one `data:` payload of the live stream. It is full state for
// the three facts it carries, not a delta, so a client that missed a frame
// recovers on the next one.
type UsageLiveFrame struct {
	// Active is the in-flight set, oldest first, so the longest-running request
	// is the first row.
	Active []UsageLiveActive `json:"active"`
	// Recent is the requests that just finished, newest first, bounded by the
	// service's read.
	Recent []UsageLiveRecent `json:"recent"`
	// ErrorProvider is the provider the gateway last reported an error for, or
	// an empty string when the last attempt to reach a provider succeeded.
	ErrorProvider string `json:"error_provider"`
}

// UsageLiveFrameFrom maps the gateway's state onto the wire frame.
//
// The order of both lists is preserved rather than re-sorted: the caller's read
// is what bounded and ordered them, and a second ordering rule here would be a
// second answer to "which request is first" that could disagree with the query.
func UsageLiveFrameFrom(markers []domain.ActiveRequest, recent []domain.UsageRecord, errorProvider string) UsageLiveFrame {
	frame := UsageLiveFrame{
		Active:        make([]UsageLiveActive, 0, len(markers)),
		Recent:        make([]UsageLiveRecent, 0, len(recent)),
		ErrorProvider: errorProvider,
	}
	for _, marker := range markers {
		frame.Active = append(frame.Active, UsageLiveActive{
			ProviderID: marker.ProviderID,
			EndpointID: marker.EndpointID,
			Model:      marker.Model,
			StartedAt:  Timestamp(marker.StartedAt),
		})
	}
	for _, record := range recent {
		frame.Recent = append(frame.Recent, UsageLiveRecent{
			RequestID:  record.RequestID(),
			ProviderID: record.ProviderID(),
			Model:      record.Model(),
			TS:         Timestamp(record.TS()),
			Status:     record.Status(),
			TokensIn:   record.TokensIn(),
			TokensOut:  record.TokensOut(),
			ErrorCode:  record.ErrorCode(),
		})
	}
	return frame
}

File: app-ui/src/lib/schemas/usage-live.ts
// The Usage live stream's frame (docs/DRAFT/010-USAGE-ENDPOINT-READINESS.md §10.3).
//
// The route this reads, `GET /api/v1/usage/live`, is not served yet (docs/DRAFT/012-USAGE-LIVE-UI-READINESS.md
// F4), so this file is the contract the panel builds against rather than a mirror of a route that exists.
// Two consequences shape it, and both are about surviving a gateway the panel does not control:
//
//   Every collection is `nullableList`, because Go marshals a nil slice as `null` rather than as `[]`.
//
//   Every field the screen does not draw is optional. A frame that carries a field this panel never reads
//   is still a frame, and rejecting it would turn a live view into an error state over a detail that does
//   not matter. Unknown keys are stripped, which is what "tolerated" means here.
//
// The frame is full state for the three fields it carries: the gateway sends the whole in-flight set, not
// a delta. `usage-live-view.ts` is where that becomes the screen's state.

import { z } from 'zod';
import { nullableList, rfc3339Timestamp, schemaRequestStatus } from './primitives';

// One request the gateway is routing right now.
//
// `started_at` is required rather than optional, and it is the only required field besides the provider:
// it is the sole input to the panel's staleness guard. Without it the panel cannot tell a request that is
// still running from one whose marker the gateway never cleared, and it would light a node forever.
export const schemaUsageLiveActive = z.object({
	provider_id: z.string().min(1),
	endpoint_id: z.string().optional(),
	model: z.string().optional(),
	started_at: rfc3339Timestamp
});

export type UsageLiveActive = z.infer<typeof schemaUsageLiveActive>;

// One request that finished recently. The gateway bounds this list (draft 010 §10.3 asks for the same
// twenty the reference keeps), so the screen renders it rather than paging it.
export const schemaUsageLiveRecent = z.object({
	request_id: z.string().min(1),
	provider_id: z.string().min(1),
	model: z.string().optional(),
	ts: rfc3339Timestamp.optional(),
	status: schemaRequestStatus.optional(),
	tokens_in: z.number().int().min(0).optional(),
	tokens_out: z.number().int().min(0).optional(),
	error_code: z.string().optional()
});

export type UsageLiveRecent = z.infer<typeof schemaUsageLiveRecent>;

export const schemaUsageLiveFrame = z.object({
	active: nullableList(schemaUsageLiveActive).optional(),
	recent: nullableList(schemaUsageLiveRecent).optional(),
	// The provider the gateway last reported an error for, or an empty string. Optional so that a frame
	// which omits it leaves the previous statement standing rather than silently clearing it.
	error_provider: z.string().optional()
});

export type UsageLiveFrame = z.infer<typeof schemaUsageLiveFrame>;
